### DISCUSSION TITLE
Add multiple HUB75 PORT .

### DISCUSSION THREAD

**Discussion by ravigyanvihar** [board707/DMD_STM32#73] [Ideas]
Sir ,
I want to make a PCB where I can add multiple hub75 port. 
Is it possible to make multiple the OE signal so at one time I can run one row of panel and so on...

**Comment by board707** [comment]
Hi
Thanks for your interest.
I doubt that STM32 boards will support more than one hub75 interface. I tested two hub75 channels on Raspberry Pico RP2040, but on stm32, I think, there are not enough resources.

**Comment by DoomHammer** [comment]
Is there an example of how to use 2 channels?

**Comment by board707** [comment]
Do you mean a two channels on  Raspberry Pico RP2040 ?
I haven't published these tests. I'll try to find my code and add it to the examples, but it will take some time.

**Comment by ravigyanvihar** [comment]
Ok sir . I am waiting

**Comment by DoomHammer** [comment]
Yes, 2 channels on Pico.

Also: how is that supposed to work. Do we get two different framebuffers or what?

**Comment by board707** [comment]
> how is that supposed to work

Sorry, I did this almost 2 years ago, in the summer of 2022, I don't remember the details. I need to find that code.

**Comment by board707** [comment]
Hi, friends !
I found my old code, but this is only a test, it is far from real use.
The general principle of operation is simple. If you look at the current version of the library, working with the hub75 channel for the RP2040 processor consists of three parts. The first part is a PIO machine that controls the multiplexer channels A B C D E. The second PIO machine outputs main RGB color channel data with CLK and LAT signals. The data for RGB pins delivered via DMA. And the third part is brightness control via PWM on the OE pin, controlled by an independent timer.

In the case of using two hub75 channels, control of the ABCDE and OE pins remains common to the two HUB channels, and two PIO machines are configured to output the color data of the first and second hub75 channels. The data of both channels is managed independently and therefore, required a two independent framebuffers.
Thus, for the operation of two parallel hub75 channels, 22 pins are required - 6 common (ABCDE + OE) and 8 for each channel - 6x RGB + LAT + CLK . Tests have shown that the scheme is quite workable, although it requires the use of the same type of panel on two channels.
As you can see, in this mode, almost all pins of the rp2040 are occupied, which limits the use of this circuit. Therefore, I did not add this mode into the library.

**Comment by board707** [comment]
I think it will be better to convert this topic to the discussion, do you agree?

**Comment by ravigyanvihar** [comment]
Yes absolutely

**Comment by ravigyanvihar** [comment]
Hi sir 
I have tried running 12 p4 80x40. ABC panels on the library on COLOR1 settings . They are running properly. 
Although there was a problem of for this setup I need to make a long cable for each row. 
I tried to edit the library in such way that each row have its own OE signal only . Remaining all pins go the same.

**Comment by ravigyanvihar** [comment]
![IMG_20240130_133543](https://github.com/board707/DMD_STM32/assets/18201517/f3b590b0-4894-43f8-b14e-bb32b1486c75)
Can we achieve this

**Comment by board707** [comment]
> I tried to edit the library in such way that each row have its own OE signal only . Remaining all pins go the same.

I'm afraid it won't work that way. Color channel pins (R1G1B1 R2G2B2) of all panels must be connected in series. 
But just the OE pin can be common to all panels.

**Comment by wsandor** [comment]
I think you can achieve this with 2 new pins and a 1:4 demultiplexer on the clock line.

**Reply by board707** [reply]
Please draw a schematic

**Reply by wsandor** [reply]
It's just an idea, unfortunately I had no time in the past months to really make and try it :
![IMG_20240202_144241](https://github.com/board707/DMD_STM32/assets/45869950/7ce0fc40-3ca2-4f00-a6b4-e01ea37620ce)
Mux1 and Mux2 is an output from a (software) counter, the counter increments when the CLK pulse count reach 1/4 of the full frame count. So every panel row gets the clock with the corresponding datas one after another. The first HUB75 gets clocks for the first 1/4 picture and so on.

**Reply by board707** [reply]
Interesting
But if I understood correctly, such a scheme does not give any gain in time. If you switch a CLK from row to row, you can only update one row at a time. That is, updating 4 rows connected according to your scheme requires exactly 4 times more time than updating one row.

The whole point of connecting panels with parallel chains is simultaneous updating, because the main limitation on the number of panels that can be connected to one controller is the update time.

Addition
I'm sorry, I forgot about the original question @ravigyanvihar asked. In his case, your option may be useful, as it allows him to connect the panels with shorter wires.

**Reply by wsandor** [reply]
Yes, you are right, my goal was also just the simpler wiring. And as a bonus, if one shift register goes wrong in the first panel for example, it corrupts only that row, not the whole display.

**Comment by ravigyanvihar** [comment]
Definately update time is a concern . 
But that's not I am trying to obtain as I am planning to make display which are bigger in P10 LED modules this total length of display is good but number of led are less giving me edge in update time .
The problem can be address in TWO ways one is different hub port for two rows or changing the chaining method .

![IMG_20240202_233446](https://github.com/board707/DMD_STM32/assets/18201517/e3042a69-aed2-4d51-9bbf-34ec26e000f8)

**Reply by board707** [reply]
> The problem can be address in TWO ways one is different hub port for two rows or changing the chaining method .

Both chain methods are supported, see Wiki: [#connecting-more-than-one-panel](https://github.com/board707/DMD_STM32/wiki/Connections---Monochrome#connecting-more-than-one-panel)

**Reply by ShiPC123** [reply]
Can P4 RGB panels be connected using this method?

**Reply by board707** [reply]
> P4 RGB panels

"P4 panels" is a very vague indication on the panel, since P4 just means that the distance between the dots on the panel is 4mm.
In general, this method should work for rgb panels too
